storm-server/src/main/java/org/apache/storm/scheduler/blacklist/BlacklistScheduler.java

```diff
@@ -132,20 +134,20 @@ private void badSupervisors(Map<String, SupervisorDetails> supervisors) {
         for (String key : badSupervisorKeys) {
             badSupervisors.put(key, cachedSupervisors.get(key));
         }
-
-        for (Map.Entry<String, SupervisorDetails> entry : supervisors.entrySet()) {
-            String key = entry.getKey();
-            SupervisorDetails supervisorDetails = entry.getValue();
-            if (cachedSupervisors.containsKey(key)) {
-                Set<Integer> badSlots = badSlots(supervisorDetails, key);
-                if (badSlots.size() > 0) { //supervisor contains bad slots
-                    badSupervisors.put(key, badSlots);
+        if (blacklistOnBadSlots) {
+            for (Map.Entry<String, SupervisorDetails> entry : supervisors.entrySet()) {
+                String key = entry.getKey();
+                SupervisorDetails supervisorDetails = entry.getValue();
+                if (cachedSupervisors.containsKey(key)) {
+                    Set<Integer> badSlots = badSlots(supervisorDetails, key);
+                    if (badSlots.size() > 0) { //supervisor contains bad slots
+                        badSupervisors.put(key, badSlots);
+                    }
+                } else {
+                    cachedSupervisors.put(key, supervisorDetails.getAllPorts()); //new supervisor to cache
```

Comment:
I think this line should execute anyway no matter if `blacklistOnBadSlots` is true or false